File: frontend/src/lib/passwordRules.js
// Must stay in sync with validate_password_strength in backend/password_rules.py.
// Purely client-side UX (live checklist) — the backend enforces the same
// rules independently and is the actual security boundary.
export const PASSWORD_RULES = [
  { id: 'length', label: 'At least 8 characters', test: (pw) => pw.length >= 8 },
  { id: 'uppercase', label: 'One uppercase letter', test: (pw) => /[A-Z]/.test(pw) },
  { id: 'lowercase', label: 'One lowercase letter', test: (pw) => /[a-z]/.test(pw) },
  { id: 'number', label: 'One number', test: (pw) => /[0-9]/.test(pw) },
  { id: 'special', label: 'One special character', test: (pw) => /[^A-Za-z0-9]/.test(pw) },
]

export function isPasswordValid(password) {
  return PASSWORD_RULES.every((rule) => rule.test(password))
}

File: backend/password_rules.py
import re

# Must stay in sync with frontend/src/lib/passwordRules.js's live checklist.
# That checklist is UX only — this is the actual security boundary.
PASSWORD_RULES = [
    ("length", "at least 8 characters", lambda pw: len(pw) >= 8),
    ("uppercase", "one uppercase letter", lambda pw: re.search(r"[A-Z]", pw) is not None),
    ("lowercase", "one lowercase letter", lambda pw: re.search(r"[a-z]", pw) is not None),
    ("number", "one number", lambda pw: re.search(r"[0-9]", pw) is not None),
    ("special", "one special character", lambda pw: re.search(r"[^A-Za-z0-9]", pw) is not None),
]


def validate_password_strength(password: str) -> None:
    """Raises ValueError listing every unmet rule, for use in a Pydantic validator."""
    unmet = [description for _, description, test in PASSWORD_RULES if not test(password)]
    if unmet:
        raise ValueError("Password must contain " + ", ".join(unmet) + ".")
